rfcs/0100-sparestir-dialect.md

```diff
@@ -0,0 +1,658 @@
+
+# Guide-level Explanation
+[guide-level-explaination]: #guide-level-explaination
+
+We have the following design goals of SparseTIR:
+
+- G0: SparseTIR is consistent with TVM's ecosystem, which means other components of TVM stack (Relax/Relay/TOPI)
+can interact with SparseTIR smoothly, and enjoy of the benefits of SparseTIR.
+- G1: SparseTIR is expressive, which means we can express most sparse operators in Deep Learning with SparseTIR.
+- G2: SparseTIR is performant, which means we can cover optimizations used in Sparse CPU/GPU libraries.
+
+We will outline the detailed design in the next section.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This section outlines the design of SparseTIR, and its interaction with existing components in TVM.
+
+## D0: Programming Interface
+
+A generic SparseTIR program looks like the following, the workload is Sampled-Dense-Dense-Matrix-Multiplication (SDDMM):
+
+```python
+@T.prim_func
+def sddmm(
+    a: T.handle,
+    b: T.handle,
+    x: T.handle,
+    y: T.handle,
+    indptr: T.handle,
+    indices: T.handle,
+    m: T.int32,
+    n: T.int32,
+    feat_size: T.int32,
+    nnz: T.int32,
+) -> None:
+    T.func_attr({"global_symbol": "main", "tir.noalias": True, "sparse_tir_level": 2})
+    # sparse axes
+    I = T.dense_fixed(m)
+    J = T.sparse_variable(I, (n, nnz), (indptr, indices), "int32")
+    J_detach = T.dense_fixed(n)
+    K = T.dense_fixed(feat_size)
+    # sparse buffers
+    A = T.match_sparse_buffer(a, (I, K), "float32")
+    B = T.match_sparse_buffer(b, (J_detach, K), "float32")
+    X = T.match_sparse_buffer(x, (I, J), "float32")
+    Y = T.match_sparse_buffer(y, (I, J), "float32")
+    # sparse iterations
+    with T.sp_iter([I, J, K], "SSR", "sddmm") as [i, j, k]:
+        with T.init():
+            Y[i, j] = 0.0
+        Y[i, j] = Y[i, j] + A[i, k] * B[j, k] * X[i, j]
+```
+
+where we have constructs like **sparse axes**,  **sparse buffers** and **sparse iterations**.
+
+### Sparse Axis
+Sparse axis is a generation of per-dimensional level formats in TACO where we annotate each dimension of a format as **dense**/**sparse** (this dimension is stored in dense or compressed storage) and **fixed**/**variable** (this dimension's extent is fixed or variable). For **sparse**/**variable** axes, we need to specify its dependent axis.
+
+- For axes that are **sparse**, we need to specify a `indices` array to store the column indices.
+- For axes that are **variable**, we need to specify an `indptr` (short for indices pointer) array to store the start offset of each row because the row length is variable and we cannot simply compute element offset with an affine map of indices.
+- An axes that is both **sparse** and **variable** need to be specified with both **indices** and **indptr** array.
+
+```python
```

Comment:
Actually, they have quite different semantics than block iterators under `T.axis` namespace. Maybe we should consider another name to avoid confusion.

Reply:
That would make sense.  Mostly my concern would be the lack of context for a first-time reader in seeing the sparse axis definitions outside of a containing namespace, and was suggesting `T.axis` as an already-existing namespace.  Maybe `T.sparse_axis.*` to avoid both the bare usage and the conflation with `T.axis` members?

Reply:
Renamed to *Sparse Iterators* to avoid confusion :)